system/doc/efficiency_guide/maps.xml

```diff
@@ -0,0 +1,647 @@
+      searched and updated regardless of the number of elements in the
+      map.</p>
+
+      <p>There is less performance to be gained by matching or
+      updating multiple elements using the map syntax on a large map
+      compared to a small map. The execution time is roughly proportional
+      to the number of elements matched or updated.</p>
+
+      <p>The storage overhead for a large map is higher than for a
+      small map. For a large map, the extra number of words besides
+      the keys and values is roughly proportional to the number of
+      elements. For a map with 33 elements the overhead is at
+      least 53 heap words according to the formula in <seeguide
+      marker="advanced#memory">Advanced</seeguide> (compared to 5
+      extra words for a small map regardless of the number of
+      elements).</p>
+
+      <p>When a large map is updated, the updated map and the original
+      map will share common parts of the HAMT, but sharing will never
+      be as effective as the sharing of the key tuple for a small
+      maps.</p>
+
+      <p>Therefore, if maps are used instead of records and it is
+      expected that many instances of the map will be created, it is
+      more efficient from a memory standpoint to avoid using large
+      maps (for example, by grouping related map elements into sub
+      maps to reduce the number of elements).</p>
+    </section>
+  </section>
+
+  <section>
+    <title>Using the Map Syntax</title>
+
+    <p>Using the map syntax is usually slightly more efficient than
+    using the corresponding function in the <seeerl
+    marker="stdlib:maps">maps</seeerl> module.</p>
+
+    <p>The gain in efficiency for the map syntax is more noticeable
+    for the following operations that can only be achieved using the
+    map syntax:</p>
+
+    <list type="bulleted">
+      <item><p>Matching multiple literal keys</p></item>
+      <item><p>Updating multiple literal keys</p></item>
+      <item><p>Adding multiple literal keys to a map</p></item>
+    </list>
+
+    <p>For example:</p>
+
+    <p><em>DO</em></p>
+    <code type="erl"><![CDATA[
+    Map = Map1#{x := X, y := Y, z := Z}]]></code>
+
+    <p><em>DO NOT</em></p>
+    <code type="erl"><![CDATA[
+    Map2 = maps:update(x, X, Map1),
+    Map3 = maps:update(y, Y, Map2),
+    Map = maps:update(z, Z, Map3)]]></code>
+
+    <p>If the map is a small map, the first example runs roughly
+    three times as fast.</p>
+
+    <p>Note that for variable keys, the elements are updated
+    sequentially from left to right. For example, given the following
+    update with variable keys:</p>
+
+    <code type="erl"><![CDATA[
+    Map = Map1#{Key1 := X, Key2 := Y, Key3 := Z}]]></code>
+
+    <p>the compiler rewrites it like this to ensure that the updates
+    are applied from left to right:</p>
+
+    <code type="erl"><![CDATA[
+    Map2 = Map1#{Key1 := X},
+    Map3 = Map2#{Key2 := Y},
+    Map = Map3#{Key3 := Z}]]></code>
```

Comment:
Not related to this document: I think this could be optimized. Instead of breaking up the update, one could check for duplicates among the keys and among duplicates of a particular key remove all that are not rightmost in the list. Is this something that is worth considering?

Reply:
I don't think it's worth it. This is an unlikely usage for a small map, and for a large map the updates will be done sequentially anyway.

Reply:
Ok, perhaps a better approach would be to have a fast path that works most of the time but can fail over to what we are doing today if there are conflicts. But maybe it is not worth it snyway if it is not a so common to use non-literal keys?